sled-agent/src/sled_agent.rs

```diff
@@ -723,11 +793,37 @@ impl SledAgent {
                 .await?;
         }
 
-        self.inner
```

Comment:
This was a big reason for the churn in this PR, and the rationale for "why did we pull up the ledger and put a lock on it?"

- The ledger covers not **only** the set of services, but also the datasets that should be used
- Whether or not we should perform updates depends on the generation number of the ledger, so we need to guard access to the ledger between "reading it" and "maybe updating it"

In the old version of this code, we "upserted filesystems" without a guard, before ensuring services. This wasn't protected by anything, and this portion of the code was not qualified by reading a generation number from the ledger.